kerckhoff/packages/models.py

```diff
@@ -84,6 +85,51 @@ def get_new_packages_from_gdrive(self) -> List["Package"]:
                 created_packages.append(package)
         return created_packages
 
+class PackageItem(models.Model):
+    TEXT = "txt"
+    AML = "aml"
+    IMAGE = "img"
+    MARKDOWN = "mdn"
+    SPREADSHEET = "xls"
+
+    DTYPE_CHOICES = (
+        (TEXT, "TEXT"),
+        (AML, "ARCHIEML"),
+        (IMAGE, "IMAGE"),
+        (MARKDOWN, "MARKDOWN"),
+        (SPREADSHEET, "SPREADSHEET"),
+    )
+
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+    # package_version = models.ForeignKey(PackageVersion, on_delete=models.CASCADE)
+    #TODO Smart way to log all PackageVersions containing this PackageItem? Optionally enable upstream changes to PackageVersions?
+    data_type = models.CharField(max_length=3, choices=DTYPE_CHOICES, default=TEXT)
+    data = JSONField(blank=True, default=dict)
+    file_name = models.CharField(max_length=64)
+    mime_types = models.CharField(max_length=64)
+
+# Snapshot of a Package instance, defined as a collection of PackageItem objects
+class PackageVersion(models.Model):
+    """
+    Snapshot of a Package instance at a particular time
+    """
+
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+    id_num = models.IntegerField(default=0) # Integer wrapper for uuid for UI referencing, set during PackageVersion creation. Zero-indexed.
+    slug = models.CharField(max_length=64, validators=[validate_slug_with_dots, ])
+    package = models.ForeignKey("Package", on_delete=models.CASCADE)
+    creator = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.PROTECT)
+    version_description = models.TextField(blank=True)
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+
+    # Stores unordered list of PackageItem foreign keys 
+    package_items = PickledObjectField(default=set)
```

Comment:
This is considered bad practice - in SQL don't store a list of foreign keys that points to other things - this breaks any sort of nice constraints on deletion or selection (you can't query/index on this field! doing a lookup based on packageversion will be very expensive). If a PackageVersion has PackageItems, then the PackageItem has a foreign key to PackageVersion - look at the one to many relationship in django https://docs.djangoproject.com/en/2.1/topics/db/examples/many_to_one/